lib/toPaths.js

```diff
@@ -207,14 +212,35 @@ function getHashCode(key) {
 }
 
 /**
- * Return true if argument is a number or can be cast to a number
+ * Return true if argument is a number or can be cast to a number.
+ * See tests for criteria.
  * @private
  */
-function isNumber(val) {
-    // parseFloat NaNs numeric-cast false positives (null|true|false|"")
-    // ...but misinterprets leading-number strings, particularly hex literals ("0x...")
-    // subtraction forces infinities to NaN
-    // adding 1 corrects loss of precision from parseFloat (#15100)
-    return !isArray(val) && (val - parseFloat(val) + 1) >= 0;
+function isSafeNumber(val) {
+    var num = val;
+    var type = typeof val;
+    if (type === typeOfString) {
+        var length = val.length;
+        // Number.MIN_SAFE_INTEGER is 17 digits including the sign.
+        // Anything longer cannot be safe.
+        if (length === 0 || length > 17) {
+            return false;
+        }
+        if (!safeNumberRegEx.test(val)) {
```

Comment:
Yup. However I'm assured that regex is fast even on slow platforms so we should be ok.